# Classical Mock Theta Functions

In his famous last letter to Hardy (1920), Ramanujan introduced **mock theta functions**
-- $q$-series that mimic theta functions in their analytic properties but are not
truly modular. He gave 17 examples organized by "order" (third, fifth, seventh).

q-Kangaroo implements all 20 classical mock theta functions. In this notebook,
we explore the third-order and fifth-order functions.

## Third-Order Mock Theta Functions

Ramanujan's third-order mock theta functions include:

$$f(q) = \sum_{n=0}^{\infty} \frac{q^{n^2}}{(-q;q)_n^2}, \qquad \phi(q) = \sum_{n=0}^{\infty} \frac{q^{n^2}}{(-q^2;q^2)_n}, \qquad \psi(q) = \sum_{n=1}^{\infty} \frac{q^{n^2}}{(q;q^2)_n}$$

In [1]:
from q_kangaroo import QSession, mock_theta_f3, mock_theta_phi3, mock_theta_psi3

s = QSession()

f3 = mock_theta_f3(s, 20)
phi3 = mock_theta_phi3(s, 20)
psi3 = mock_theta_psi3(s, 20)

print("f(q) =", f3)
print("phi(q) =", phi3)
print("psi(q) =", psi3)

## Non-Negative Coefficients of $\psi(q)$

The third-order $\psi(q) = \sum_{n=1}^{\infty} \frac{q^{n^2}}{(q;q^2)_n}$ has
a combinatorial interpretation: its coefficients count certain partition-like objects.
In particular, all coefficients are **non-negative** and **non-decreasing** for large $n$:

In [2]:
# Compute psi(q) to 30 terms -- all coefficients are positive
psi3_30 = mock_theta_psi3(s, 30)
psi3_30

q + q^2 + q^3 + 2*q^4 + 2*q^5 + 2*q^6 + 3*q^7 + 3*q^8 + 4*q^9 + 5*q^10 + 5*q^11 + 6*q^12 + 7*q^13 + 8*q^14 + 9*q^15 + 11*q^16 + 12*q^17 + 13*q^18 + 16*q^19 + 17*q^20 + 19*q^21 + 22*q^22 + 24*q^23 + 27*q^24 + 31*q^25 + 34*q^26 + 37*q^27 + 42*q^28 + 46*q^29 + O(q^30)

Every coefficient is positive: $1, 1, 1, 2, 2, 2, 3, 3, 4, 5, 5, 6, 7, 8, 9, 11, \ldots$
The growth rate is related to $\exp(\pi\sqrt{n/3}) / (4n)$, characteristic of
mock theta functions.

## All Seven Third-Order Mock Theta Functions

Ramanujan defined 7 third-order mock theta functions: $f$, $\phi$, $\psi$, $\chi$, $\omega$, $\nu$, $\rho$. We have already seen $f$, $\phi$, and $\psi$ above. Let us compute the remaining four:

$$\chi(q) = \sum_{n=0}^{\infty} \frac{q^{n^2}}{\prod_{k=1}^{n}(1 - q^k + q^{2k})}, \qquad \omega(q) = \sum_{n=0}^{\infty} \frac{q^{2n(n+1)}}{(q;q^2)_{n+1}^2}$$

$$\nu(q) = \sum_{n=0}^{\infty} \frac{q^{n(n+1)}}{(-q;q^2)_{n+1}}, \qquad \rho(q) = \sum_{n=0}^{\infty} \frac{q^{2n(n+1)}}{\prod_{k=0}^{n}(1 + q^{2k+1} + q^{4k+2})}$$

In [3]:
from q_kangaroo import mock_theta_chi3, mock_theta_omega3, mock_theta_nu3, mock_theta_rho3

chi3 = mock_theta_chi3(s, 20)
omega3 = mock_theta_omega3(s, 20)
nu3 = mock_theta_nu3(s, 20)
rho3 = mock_theta_rho3(s, 20)

print("chi(q) =", chi3)
print("omega(q) =", omega3)
print("nu(q) =", nu3)
print("rho(q) =", rho3)

Notice the variety of coefficient patterns:

- $\chi(q)$ is **sparse** with mostly 0/1 coefficients and occasional $-1$ terms. The cyclotomic factors $(1 - q^k + q^{2k})$ in its denominator impose strong cancellation.
- $\omega(q)$ has **rapidly growing** all-positive coefficients: $1, 2, 3, 4, 6, 8, 10, 14, 18, \ldots$
- $\nu(q)$ has **alternating signs**, reflecting the $(-q; q^2)$ factors in its denominator.
- $\rho(q)$ is sparse like $\chi$, with cyclotomic factors $(1 + q^{2k+1} + q^{4k+2})$.

The seven third-order functions satisfy classical linear relations over theta functions. For instance, Watson (1936) proved relations connecting $f$, $\phi$, $\psi$, $\chi$, $\omega$, $\nu$, and $\rho$ to each other via products of theta functions.

## Fifth-Order Mock Theta Functions

Ramanujan also defined 10 fifth-order mock theta functions. Here are three of them:

In [4]:
from q_kangaroo import mock_theta_f0_5, mock_theta_f1_5, mock_theta_phi0_5

f05 = mock_theta_f0_5(s, 20)
f15 = mock_theta_f1_5(s, 20)
phi05 = mock_theta_phi0_5(s, 20)

print("f0(q) =", f05)
print("f1(q) =", f15)
print("phi0(q) =", phi05)

## Seventh-Order Mock Theta Functions

Ramanujan also listed 3 seventh-order mock theta functions: $F_0$, $F_1$, $F_2$. These are the most computationally intensive, as each term requires a fresh Pochhammer product computation (the denominator base shifts with $n$, preventing incremental reuse).

$$F_0(q) = \sum_{n=0}^{\infty} \frac{q^{n^2}}{(q^{n+1};q)_n}, \qquad F_1(q) = \sum_{n=0}^{\infty} \frac{q^{n^2}}{(q^n;q)_n}, \qquad F_2(q) = \sum_{n=0}^{\infty} \frac{q^{n^2+n}}{(q^{n+1};q)_{n+1}}$$

In [5]:
from q_kangaroo import mock_theta_cap_f0_7, mock_theta_cap_f1_7, mock_theta_cap_f2_7

F0_7 = mock_theta_cap_f0_7(s, 15)
F1_7 = mock_theta_cap_f1_7(s, 15)
F2_7 = mock_theta_cap_f2_7(s, 15)

print("F0(q) =", F0_7)
print("F1(q) =", F1_7)
print("F2(q) =", F2_7)

The seventh-order functions were the most mysterious of Ramanujan's original list. All three have non-negative coefficients and moderate growth. Their modular properties were only fully understood after Zwegers' work in 2002, which placed them within the framework of harmonic Maass forms.

## Appell-Lerch Sums

An **Appell-Lerch sum** is a bilateral series central to the modern theory of mock theta functions:

$$m(x, q, z) = \frac{1}{j(z;q)} \sum_{r=-\infty}^{\infty} \frac{(-1)^r q^{r(r-1)/2} z^r}{1 - x q^{r-1} z}$$

where $j(z;q) = (z;q)_\infty(q/z;q)_\infty(q;q)_\infty$ is the Jacobi theta product. Zwegers showed that **every mock theta function** can be expressed in terms of Appell-Lerch sums.

For integer specializations $x = q^a$, $z = q^b$, the Jacobi theta $j(q^b; q)$ vanishes (since one of the infinite products always has a factor $(1-q^0) = 0$). The `appell_lerch_m` function computes the **raw bilateral sum** (without the $j(z;q)$ normalization), which is well-defined and useful for verifying identities where the $j$ factors cancel.

In [6]:
from q_kangaroo import appell_lerch_m

# Compute the bilateral sum with a = q^2, z = q
al1 = appell_lerch_m(s, 2, 1, 20)
print("S(q^2, q, q) =", al1)

In [7]:
# Different specialization: a = q^3, z = q
al2 = appell_lerch_m(s, 3, 1, 20)
print("S(q^3, q, q) =", al2)

The extreme sparsity of these bilateral sums for integer parameters is a consequence of massive cancellation in the bilateral series. Each positive-index term $r > 0$ nearly cancels with its negative-index counterpart $-r$, leaving only isolated contributions. This cancellation is intimately related to modular properties: the Appell-Lerch sums satisfy beautiful transformation properties under $q \to q^n$, and different specializations produce different mock theta functions.

The coefficient $3q^3$ for $(a, z) = (q^2, q)$ and $-4q^6$ for $(a, z) = (q^3, q)$ are exact residual values after bilateral cancellation.

## Universal Mock Theta Functions $g_2$ and $g_3$

The **universal mock theta functions** $g_2(x, q)$ and $g_3(x, q)$ were introduced by Gordon and McIntosh (2000). They are called "universal" because **all classical mock theta functions** can be expressed in terms of them.

$$g_2(x, q) = x^{-1}(-q; q)_\infty \sum_{n=0}^{\infty} \frac{(-q)^{n^2}}{(x; q^2)_{n+1}(q^2/x; q^2)_{n+1}}$$

$$g_3(x, q) = \sum_{n=0}^{\infty} \frac{q^{n(n+1)}}{(x; q)_{n+1}(q/x; q)_{n+1}}$$

For integer powers $x = q^a$ with $a \geq 2$, the denominators $(q/x; q)_{n+1} = (q^{1-a}; q)_{n+1}$ degenerate for $n \geq a-1$, so only finitely many terms contribute. The implementations handle this algebraically by rewriting the negative-exponent factors.

In [8]:
from q_kangaroo import universal_mock_theta_g2

# Compute g2(q^3, q) -- the universal mock theta function
g2 = universal_mock_theta_g2(s, 3, 20)
g2

-q^-1 - 1 - q - q^2 + q^3 + 4*q^4 + 10*q^5 + 21*q^6 + 38*q^7 + 63*q^8 + 102*q^9 + 157*q^10 + 234*q^11 + 342*q^12 + 488*q^13 + 683*q^14 + 944*q^15 + 1285*q^16 + O(q^17)

In [9]:
from q_kangaroo import universal_mock_theta_g3

# Compute g3(q^3, q) -- companion universal mock theta function
g3 = universal_mock_theta_g3(s, 3, 20)
print("g3(q^3, q) =", g3)

Comparing $g_2$ and $g_3$ at $x = q^3$:

- $g_2$ has a pole at $q^{-1}$ and extremely rapid growth ($1285$ at $q^{16}$). It involves the base $q^2$ through $(x; q^2)_{n+1}$.
- $g_3$ starts at $-q^2$ and grows more moderately ($51$ at $q^{19}$). It uses base $q$ through $(x; q)_{n+1}$.

Both are meromorphic functions with poles along $x \in q^{\mathbb{Z}}$. The universality theorem states that for any classical mock theta function $\mu(q)$, there exist parameters $(x_1, \ldots, x_k)$ and theta function coefficients such that $\mu(q) = \sum_i c_i \cdot g_j(x_i, q) + \text{theta functions}$ for $j \in \{2, 3\}$.

## Comparing Orders: Coefficient Growth

Mock theta functions of different orders exhibit different coefficient growth patterns. Third-order functions (like $f$) grow as $\sim \exp(\pi\sqrt{n/3})/(4n)$, while fifth-order functions tend to have smaller coefficients at the same index. Let us compare representative functions from each order:

In [10]:
# Compare f3, f0_5, and F0_7 side by side to order 15
f3_15 = mock_theta_f3(s, 15)
f05_15 = mock_theta_f0_5(s, 15)
F07_15 = mock_theta_cap_f0_7(s, 15)

print("Third-order f(q): ", f3_15)
print("Fifth-order f0(q):", f05_15)
print("Seventh-order F0(q):", F07_15)

Key observations:

- **Third-order $f(q)$** has the largest absolute values ($|-17|$ at $q^{14}$) and alternating signs. The alternation comes from the $(-q;q)_n^2$ denominator.
- **Fifth-order $f_0(q)$** is sparser, with many zero coefficients and smaller absolute values. The maximum is $|-2|$ through order 14.
- **Seventh-order $F_0(q)$** has all non-negative coefficients (no alternation since the denominator $(q^{n+1};q)_n$ has no sign changes) and moderate growth.

The relationship between "order" and growth reflects the level of the associated modular form: third-order mock theta functions relate to weight $1/2$ forms on $\Gamma_0(144)$, while higher-order functions live on congruence subgroups of higher level.

## Connection to Zwegers' Thesis

Zwegers' 2002 thesis revolutionized the understanding of mock theta functions by showing
that Ramanujan's mock theta functions are the **holomorphic parts** of weight $1/2$
harmonic Maass forms. The key insight is that the Appell-Lerch sums $m(x,q,z)$ can be
"completed" by adding a non-holomorphic correction term involving an error function integral.
The completion transforms mock theta functions into genuine harmonic Maass forms of weight $1/2$.

This framework:

1. **Explains** why mock theta functions almost-but-not-quite satisfy modular transformation laws: the non-holomorphic correction (an Eichler integral of a unary theta function) accounts for the "error of modularity."

2. **Provides the error of modularity explicitly** via Eichler integrals: $\widehat{\mu}(\tau) = \mu(\tau) + R(\tau)$ where $R(\tau)$ is a non-holomorphic period integral of a weight $3/2$ unary theta function.

3. **Connects to indefinite theta functions and quantum modular forms** (Zagier): the "quantum" modularity of $\mu$ at rationals is governed by the asymptotic behavior of the period integral $R$.

4. **Has far-reaching applications**: black hole entropy in string theory (via wall-crossing formulas), umbral moonshine (connecting mock modular forms to sporadic groups), quantum invariants of 3-manifolds, and combinatorial partition statistics.

q-Kangaroo provides computational tools for exploring these connections through
exact symbolic computation of the relevant $q$-series: all 20 classical mock theta functions,
the Appell-Lerch bilateral sums, and the universal mock theta functions $g_2$ and $g_3$
of Gordon-McIntosh.